clients/src/main/java/org/apache/kafka/common/utils/CircularIterator.java

```diff
@@ -14,36 +14,69 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 package org.apache.kafka.common.utils;
 
+import java.util.Collection;
 import java.util.Iterator;
-import java.util.List;
+import java.util.Objects;
 
+/**
+ * An iterator that cycles through the Iterator of a Collection indefinitely.
+ * Useful for tasks such as round-robin load balancing.
+ */
 public class CircularIterator<T> implements Iterator<T> {
-    int i = 0;
-    private List<T> list;
 
-    public CircularIterator(List<T> list) {
-        if (list.isEmpty()) {
+    private final Iterable<T> iterable;
+    private Iterator<T> iterator;
+    private T peek;
+
+    /**
+     * Create a new instance of a CircularIterator. The ordering of this
+     * Iterator will be dictated by the Iterator returned by Collection itself.
+     *
+     * @param col The collection to iterate indefinitely
+     *
+     * @throws NullPointerException if col is {@code null}
+     * @throws IllegalArgumentException if col is empty.
+     */
+    public CircularIterator(final Collection<T> col) {
+        this.iterable = Objects.requireNonNull(col);
+        this.iterator = col.iterator();
+        this.peek = null;
+        if (col.isEmpty()) {
             throw new IllegalArgumentException("CircularIterator can only be used on non-empty lists");
         }
-        this.list = list;
     }
 
     @Override
     public boolean hasNext() {
-        return true;
+        if (this.peek != null) {
+            return true;
+        }
+        if (!this.iterator.hasNext()) {
+            this.iterator = this.iterable.iterator();
+        }
+        return this.iterator.hasNext();
     }
 
     @Override
     public T next() {
-        T next = list.get(i);
-        i = (i + 1) % list.size();
-        return next;
+        final T nextValue;
+        if (this.peek != null) {
+            nextValue = this.peek;
+            this.peek = null;
+        } else {
+            nextValue = this.iterator.next();
+        }
+        return nextValue;
     }
 
     public T peek() {
-        return list.get(i);
+        if (this.peek == null) {
```

Comment:
Need to add a test with a collection containing `null` -- this logic will break.  Best to always have the peek value, initializing it in the constructor.